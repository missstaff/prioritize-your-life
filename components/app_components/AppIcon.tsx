import React from "react";
import { View } from "react-native";
import { ScaledSheet, s } from "react-native-size-matters";
import Ionicons from "@expo/vector-icons/Ionicons";
import { type IconProps } from "@expo/vector-icons/build/createIconSet";
import { type ComponentProps } from "react";

/**
 * Renders an app icon component.
 *
 * @param style - The style object for the component.
 * @param rest - Additional props for the component.
 * @returns The rendered app icon component.
 */
export function AppIcon({
  style,
  ...rest
}: IconProps<ComponentProps<typeof Ionicons>["name"] | any>): JSX.Element {
  return (
    <View style={[styles.container, style]}>
      <Ionicons size={s(28)} {...rest} />
    </View>
  );
}

const styles = ScaledSheet.create({
  container: {
    alignItems: "center",
    backgroundColor: "transparent",
    borderRadius: 100,
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    padding: s(2.5),
  },
});
