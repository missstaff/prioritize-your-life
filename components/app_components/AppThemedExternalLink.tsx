import { Href, Link } from "expo-router";
import { Platform } from "react-native";
import { openBrowserAsync } from "expo-web-browser";
import { type ComponentProps } from "react";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORTHEME } from "@/constants/Colors";

type Props = Omit<ComponentProps<typeof Link>, "href"> & {
  href: Href<string | object>;
};

const AppThemedExternalLink = ({ href, style, ...rest }: Props) => {
  const color = useThemeColor(
    { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
    "background"
  );
  return (
    <Link
      target="_blank"
      {...rest}
      href={href}
      onPress={async (event) => {
        if (Platform.OS !== "web") {
          // Prevent the default behavior of linking to the default browser on native.
          event.preventDefault();
          // Open the link in an in-app browser.
          await openBrowserAsync(href as string);
        }
      }}
      style={[{ color }, style]}
    />
  );
};

export default AppThemedExternalLink;
