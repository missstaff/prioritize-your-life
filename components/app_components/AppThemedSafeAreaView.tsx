import { COLORTHEME } from "@/constants/Colors";
import { useThemeColor } from "@/hooks/useThemeColor";
import { SafeAreaView } from "react-native";


const AppThemedSafeAreaView  = ({ children }: { children: React.ReactNode }) => {
    const backgroundColor = useThemeColor(
        { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
        "background"
      );
    return (
        <SafeAreaView style={{flex: 1, backgroundColor: backgroundColor}}>
            {children}
        </SafeAreaView>
    );
};
export default AppThemedSafeAreaView;