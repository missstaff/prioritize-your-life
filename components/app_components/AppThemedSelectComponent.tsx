
const AppThemedSelectComponent = () => {};

export default AppThemedSelectComponent;