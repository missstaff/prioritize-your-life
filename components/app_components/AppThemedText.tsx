import { Text, type TextProps } from "react-native";
import { ScaledSheet, s } from "react-native-size-matters";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORS, COLORTHEME } from "@/constants/Colors";

export type ThemedTextProps = TextProps & {
  type?: "default" | "title" | "defaultSemiBold" | "subtitle" | "link";
};

const AppThemedText = ({
  style,
  type = "default",
  ...rest
}: ThemedTextProps) => {
  const color = useThemeColor(
    { light: COLORTHEME.light.text, dark: COLORTHEME.dark.text },
    "text"
  );

  return (
    <Text
      style={[
        { color },
        type === "default" ? styles.default : undefined,
        type === "title" ? styles.title : undefined,
        type === "defaultSemiBold" ? styles.defaultSemiBold : undefined,
        type === "subtitle" ? styles.subtitle : undefined,
        type === "link" ? styles.link : undefined,
        style,
      ]}
      {...rest}
    />
  );
};

const styles = ScaledSheet.create({
  default: {
    fontSize: s(16),
    lineHeight: s(24),
  },
  defaultSemiBold: {
    fontSize: s(16),
    fontWeight: "600",
    lineHeight: s(24),
  },
  title: {
    fontSize: s(32),
    fontWeight: "bold",
    lineHeight: s(32),
  },
  subtitle: {
    fontSize: s(20),
    fontWeight: "bold",
  },
  link: {
    color: COLORS.primary,
    fontSize: s(16),
    lineHeight: s(30),
  },
});

export default AppThemedText;
