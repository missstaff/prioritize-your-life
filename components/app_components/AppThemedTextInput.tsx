import React, { useState } from "react";
import {
  StyleProp,
  TextInput,
  TextStyle,
  TouchableOpacity,
  ViewStyle,
  Platform,
} from "react-native";
import { DateTimePickerAndroid } from "@react-native-community/datetimepicker";
import { ScaledSheet, s } from "react-native-size-matters";
import AppIcon from "./AppIcon";
import AppThemedView from "./AppThemedView";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORTHEME } from "@/constants/Colors";
import { TransactionState } from "@/store/transaction/transaction-reducer";

interface AppThemedTextInputProps {
  containerStyle?: StyleProp<ViewStyle>;
  data?: any[] | undefined;
  iconName?: string;
  inputStyle?: StyleProp<TextStyle>;
  keyboardType?: "default" | "numeric";
  placeholder: string;
  secureEntry: boolean;
  value: string;
  checkValue: (value: string) => void;
  setValue: (value: string) => void;
}

const AppThemedTextInput = ({
  containerStyle,
  data,
  inputStyle,
  iconName,
  keyboardType,
  placeholder,
  secureEntry,
  value,
  checkValue,
  setValue,
  ...rest
}: AppThemedTextInputProps) => {
  const textColor = useThemeColor({}, "text");
  let backgroundColor = useThemeColor(
    { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
    "background"
  );

  const [isPasswordVisible, setIsPasswordVisible] = useState(!secureEntry);

  const handleCalendarPress = () => {
    if (Platform.OS === "android") {
      DateTimePickerAndroid.open({
        mode: "date",
        value: new Date(),
        onChange: (event, selectedDate) => {
          if (selectedDate) {
            setValue(formatDate(selectedDate));
          }
        },
      });
    }
  };

  return (
    <AppThemedView style={[styles.inputContainer, containerStyle]}>
      <TextInput
        keyboardType={keyboardType}
        onBlur={(e) => {
          if (value.length > 0) checkValue(value);
        }}
        onChangeText={(text) => {
          setValue(text);
        }}
        placeholder={placeholder}
        placeholderTextColor="#999"
        secureTextEntry={!isPasswordVisible}
        style={[
          styles.input,
          { backgroundColor: backgroundColor, color: textColor },
          inputStyle,
        ]}
        defaultValue={value}
        value={value}
        {...rest}
      />
      {secureEntry && (
        <TouchableOpacity
          onPress={() => setIsPasswordVisible(!isPasswordVisible)}
          testID="passwordVisibilityToggle"
        >
          <AppIcon
            name={!isPasswordVisible ? "eye-off" : "eye"}
            size={s(24)}
            color="#ccc"
          />
        </TouchableOpacity>
      )}

      {iconName && (
        <TouchableOpacity
          disabled={iconName !== "calendar"}
          onPress={() => iconName === "calendar" && handleCalendarPress()}
        >
          <AppIcon name={iconName} size={s(24)} color="#ccc" />
        </TouchableOpacity>
      )}
    </AppThemedView>
  );
};

const styles = ScaledSheet.create({
  inputContainer: {
    alignItems: "center",
    borderColor: "#ccc",
    borderRadius: s(5),
    borderWidth: s(1),
    flexDirection: "row",
    marginBottom: s(10),
    paddingHorizontal: s(10),
    width: "80%",
  },
  input: {
    flex: 1,
    padding: s(10),
  },
});

const formatDate = (date: Date): string => {
  const day = String(date.getDate()).padStart(2, "0");
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const year = date.getFullYear();
  return `${month}/${day}/${year}`;
};

export default AppThemedTextInput;
