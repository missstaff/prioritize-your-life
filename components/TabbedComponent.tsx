import { Pressable, View } from "react-native";
import { ScaledSheet, s, vs } from "react-native-size-matters";
import AppThemedText from "@/components/app_components/AppThemedText";
import { COLORS } from "@/constants/Colors";

interface TabbedComponentProps {
  children: React.ReactNode[];
  selectedTab: number;
  tabs: string[];
  setSelectedTab: (selectedTab: number) => void;
}

const TabbedComponent: React.FC<TabbedComponentProps> = ({
  children,
  selectedTab,
  setSelectedTab,
  tabs,
}) => {
  return (
    <>
      <View style={styles.tabContainer}>
        {tabs.map((tab, index) => (
          <Pressable key={index} onPress={() => setSelectedTab(index)}>
            <AppThemedText
              style={[selectedTab === index && styles.selectedTab, styles.tab]}
            >
              {tab}
            </AppThemedText>
          </Pressable>
        ))}
      </View>
      {children[selectedTab]}
    </>
  );
};

const styles = ScaledSheet.create({
  tabContainer: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "center",
    paddingBottom: vs(20),
    width: "100%",
  },
  tab: {
    paddingHorizontal: s(15),
  },
  selectedTab: {
    color: COLORS.primary,
  },
});

export default TabbedComponent;
