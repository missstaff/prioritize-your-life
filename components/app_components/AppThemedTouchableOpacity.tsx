import React from "react";
import { ScaledSheet, s } from "react-native-size-matters";
import {
  StyleProp,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
} from "react-native";
import { AppThemedText } from "./AppThemedText";
import { COLORS } from "@/constants/Colors";

export type ThemedAppThemedTouchableOpacity = TouchableOpacityProps & {
  children: React.ReactNode;
  textStyles?: StyleProp<TextStyle>;
  btnStyles?: StyleProp<ViewStyle>;
  onPress?: () => void;
};

/**
 * Custom TouchableOpacity component for the app.
 * @param {TouchableOpacityProps} props - The TouchableOpacity props.
 * @param {React.ReactNode} props.children - The children to be rendered inside the TouchableOpacity.
 * @param {StyleProp<ViewStyle>} props.textStyles - The styles to be applied to the text inside the TouchableOpacity.
 * @param {StyleProp<ViewStyle>} props.btnStyles - The styles to be applied to the TouchableOpacity.
 * @param {Function} props.onPress - The function to be called when the TouchableOpacity is pressed.
 * @param {boolean} props.disabled - A boolean value to disable the TouchableOpacity.
 * @returns {JSX.Element} - The rendered TouchableOpacity component.
 */
const AppThemedTouchableOpacity = ({
  children,
  onPress,
  textStyles,
  btnStyles,
  disabled
}: ThemedAppThemedTouchableOpacity): JSX.Element => {
  return (
    <TouchableOpacity disabled={disabled} onPress={onPress} style={[styles.button, btnStyles]}>
      <AppThemedText style={[styles.buttonText, textStyles]}>
        {children}
      </AppThemedText>
    </TouchableOpacity>
  );
};

export default AppThemedTouchableOpacity;

const styles = ScaledSheet.create({
  button: {
    alignItems: "center",
    backgroundColor: COLORS.primary,
    borderRadius: s(5),
    marginBottom: s(10),
    padding: s(10),
    width: "80%",
  },
  buttonText: {
    color: COLORS.white,
  },
});
