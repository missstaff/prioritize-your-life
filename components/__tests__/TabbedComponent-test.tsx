import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import TabbedComponent from "@/components/TabbedComponent";
import { AppThemedText } from "@/components/app_components/AppThemedText";

describe("TabbedComponent", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const mockSetSelectedTab = jest.fn();
  const tabs = ["Tab 1", "Tab 2", "Tab 3"];
  const children = [
    <AppThemedText key="0">Content for Tab 1</AppThemedText>,
    <AppThemedText key="1">Content for Tab 2</AppThemedText>,
    <AppThemedText key="2">Content for Tab 3</AppThemedText>,
  ];

  it("renders correctly", () => {
    const tree = renderer
      .create(
        <TabbedComponent
          selectedTab={0}
          setSelectedTab={mockSetSelectedTab}
          tabs={tabs}
        >
          {children}
        </TabbedComponent>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with given tabs", () => {
    const { getByText } = render(
      <TabbedComponent
        selectedTab={0}
        setSelectedTab={mockSetSelectedTab}
        tabs={tabs}
      >
        {children}
      </TabbedComponent>
    );

    tabs.forEach((tab) => {
      expect(getByText(tab)).toBeTruthy();
    });
  });

  it("calls setSelectedTab when a tab is pressed", () => {
    const { getByText } = render(
      <TabbedComponent
        selectedTab={0}
        setSelectedTab={mockSetSelectedTab}
        tabs={tabs}
      >
        {children}
      </TabbedComponent>
    );

    fireEvent.press(getByText("Tab 2"));
    expect(mockSetSelectedTab).toHaveBeenCalledWith(1);

    fireEvent.press(getByText("Tab 3"));
    expect(mockSetSelectedTab).toHaveBeenCalledWith(2);
  });

  it("displays the correct content for the selected tab", () => {
    const { getByText, rerender } = render(
      <TabbedComponent
        selectedTab={0}
        setSelectedTab={mockSetSelectedTab}
        tabs={tabs}
      >
        {children}
      </TabbedComponent>
    );

    expect(getByText("Content for Tab 1")).toBeTruthy();

    rerender(
      <TabbedComponent
        selectedTab={1}
        setSelectedTab={mockSetSelectedTab}
        tabs={tabs}
      >
        {children}
      </TabbedComponent>
    );

    expect(getByText("Content for Tab 2")).toBeTruthy();

    rerender(
      <TabbedComponent
        selectedTab={2}
        setSelectedTab={mockSetSelectedTab}
        tabs={tabs}
      >
        {children}
      </TabbedComponent>
    );

    expect(getByText("Content for Tab 3")).toBeTruthy();
  });
});
