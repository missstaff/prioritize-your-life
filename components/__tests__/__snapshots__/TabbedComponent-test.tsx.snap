// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabbedComponent renders correctly 1`] = `
[
  <View
    style={
      [
        {
          "backgroundColor": "#151718",
        },
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 39.23529411764706,
          "width": "100%",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          [
            {
              "color": "#ECEDEE",
            },
            {
              "fontSize": 34.285714285714285,
              "lineHeight": 51.42857142857143,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            [
              {
                "color": "#CD03EF",
              },
              {
                "paddingHorizontal": 32.14285714285714,
              },
            ],
          ]
        }
      >
        Tab 1
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          [
            {
              "color": "#ECEDEE",
            },
            {
              "fontSize": 34.285714285714285,
              "lineHeight": 51.42857142857143,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            [
              false,
              {
                "paddingHorizontal": 32.14285714285714,
              },
            ],
          ]
        }
      >
        Tab 2
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          [
            {
              "color": "#ECEDEE",
            },
            {
              "fontSize": 34.285714285714285,
              "lineHeight": 51.42857142857143,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            [
              false,
              {
                "paddingHorizontal": 32.14285714285714,
              },
            ],
          ]
        }
      >
        Tab 3
      </Text>
    </View>
  </View>,
  <Text
    style={
      [
        {
          "color": "#ECEDEE",
        },
        {
          "fontSize": 34.285714285714285,
          "lineHeight": 51.42857142857143,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Content for Tab 1
  </Text>,
]
`;
